server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -318,7 +320,8 @@ public ShardFailedClusterStateTaskExecutor(AllocationService allocationService,
         @Override
         public ClusterState execute(
             ClusterState currentState,
-            List<ClusterStateTaskExecutor.TaskContext<FailedShardUpdateTask>> taskContexts
+            List<TaskContext<FailedShardUpdateTask>> taskContexts,
+            Supplier<Releasable> dropHeadersContextSupplier
```

Comment:
It looks like this is not required in every implementation. It would also be great to minimize the amount of parameters.
I wonder if there is a nice way to convert it to a method call within the interface?

Reply:
Not really. You can do this:

```java
    default ClusterState execute(ClusterState currentState, List<TaskContext<T>> taskContexts, Supplier<Releasable> dropHeadersContextSupplier) throws Exception {
        return execute(currentState, taskContexts);
    }

    ClusterState execute(ClusterState currentState, List<TaskContext<T>> taskContexts) throws Exception;
```

I don't like this pattern because the implementations that _do_ need the supplier must also supply a dummy implementation of the 2-arg method which is never called. I'd rather just have the extra arg throughout.

Reply:
I think it would be nice to introduce an object for the parameters, like `BatchContext`. That would have `state` and `taskContexts` methods as well as a `dropHeaders` method. This would remove the high visibility of this extra arg as well as make the signature nicer (no need for `Supplier`). It would allow future similar extensions with less effect across code base. And avoid the multi-line breaking of the formatter (minor point).

Reply:
:+1: makes sense. I'll not do that here but I'll open a follow-up.

Reply:
Actually I'll do that first, that'll reduce the noise from this change.